scripts/spec_xml/generate_spec_xml.py

```diff
@@ -15,21 +15,167 @@
 # limitations under the License.
 
 import glob
+import logging
 import os
 import re
 import subprocess
+import sys
+from dataclasses import dataclass, field
+from typing import Optional
 
 import click
 
 DEFAULT_CHIP_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
 DEFAULT_OUTPUT_DIR = os.path.abspath(os.path.join(DEFAULT_CHIP_ROOT, 'data_model'))
 
 
+# Known TOC differences
+# - NOC has an additional command section
+# - DiagnosticsGeneral has an additional DoNotUse attribute that does not appear in the TOC
+# - GeneralCommissioning has an additional section in the TOC for common error handling
+# - ballast configuration is missing two attributes from the TOC (D attributes)
+# - LevelControl has an additional attribute in the TOC for scene table extensions
+# - ModeSelect has an attitional attribute in the TOC for scene table extensions
+# - Scenes has an extra command in the TOC for usage notes
+# - onoff has an additional attribute in the TOC for scene table extensions
+# - color control is all over the place - ignore this for attributes.
+# - microwave oven has an additional TOC section in commands for operational state description
+# - EnergyEVSE - attributes SessionID, SessionDuration, SessionEnergyCharged and SessionEnergyDischarged are all collapsed into one section, meaning the TOC has three fewer sections than the XML
+
+
+# Known ok warnings
+# - ballast config has two D attributes (power on level and power on fade time) that don't have access
+
+@dataclass
+class ElementCounts:
+    attributes: int
+    commands: int
+    events: int
+
+
 def get_xml_path(filename, output_dir):
     xml = os.path.basename(filename).replace('.adoc', '.xml')
     return os.path.abspath(os.path.join(output_dir, xml))
 
 
+def get_toc(filename: str, dry_run: bool) -> Optional[str]:
+    # -o - sends output to stdout so we can scrape it without needing an intermediate file
+    cmd = ['asciidoctor', '-a', 'toc', '-a', 'toclevel=2', '-d', 'book', '-o', '-', filename]
+    if dry_run:
+        print(cmd)
+        return
+    try:
+        process = subprocess.run(cmd, capture_output=True)
```

Comment:
how abouthttps://docs.python.org/3/library/subprocess.html#subprocess.getstatusoutput ? Encoding should probably be utf8.